osc/osc_test.go

```diff
@@ -111,53 +113,80 @@ func TestAddMsgHandlerWithInvalidAddress(t *testing.T) {
 	}
 }
 
+// These tests stop the server by forcibly closing the connection, which causes
+// a "use of closed network connection" error the next time we try to read from
+// the connection. As a workaround, this wraps server.ListenAndServe() in an
+// error-handling layer that doesn't consider "use of closed network connection"
+// an error.
+//
+// Open question: is this desired behavior, or should server.serve return
+// successfully in cases where it would otherwise throw this error?
+func serveUntilInterrupted(server *Server) error {
+	if err := server.ListenAndServe(); err != nil &&
+		!strings.Contains(err.Error(), "use of closed network connection") {
+		return err
+	}
+
+	return nil
+}
+
 func TestServerMessageDispatching(t *testing.T) {
 	finish := make(chan bool)
 	start := make(chan bool)
 	done := sync.WaitGroup{}
 	done.Add(2)
 
-	// Start the OSC server in a new go-routine
-	go func() {
-		conn, err := net.ListenPacket("udp", "localhost:6677")
-		if err != nil {
-			t.Fatal(err)
-		}
-		defer conn.Close()
+	port := 6677
+	addr := "localhost:" + strconv.Itoa(port)
+
+	server := &Server{Addr: addr, Dispatcher: NewStandardDispatcher()}
+	defer server.CloseConnection()
+
+	if err := server.Dispatcher.(*StandardDispatcher).AddMsgHandler(
+		"/address/test",
+		func(msg *Message) {
+			defer func() {
+				server.CloseConnection()
+				finish <- true
+			}()
 
-		d := NewStandardDispatcher()
-		err = d.AddMsgHandler("/address/test", func(msg *Message) {
 			if len(msg.Arguments) != 1 {
 				t.Error("Argument length should be 1 and is: " + string(len(msg.Arguments)))
 			}
 
 			if msg.Arguments[0].(int32) != 1122 {
 				t.Error("Argument should be 1122 and is: " + string(msg.Arguments[0].(int32)))
 			}
+		},
+	); err != nil {
+		t.Error("Error adding message handler")
+	}
 
-			// Stop OSC server
-			conn.Close()
-			finish <- true
-		})
-		if err != nil {
-			t.Error("Error adding message handler")
-		}
-
-		server := &Server{Addr: "localhost:6677", Dispatcher: d}
+	// Server goroutine
+	go func() {
 		start <- true
-		server.Serve(conn)
+
+		if err := serveUntilInterrupted(server); err != nil {
+			t.Errorf("error during Serve: %s", err.Error())
+		}
 	}()
 
+	// Client goroutine
 	go func() {
 		timeout := time.After(5 * time.Second)
 		select {
 		case <-timeout:
 		case <-start:
 			time.Sleep(500 * time.Millisecond)
-			client := NewClient("localhost", 6677)
+			client := NewClient("localhost", port)
 			msg := NewMessage("/address/test")
 			msg.Append(int32(1122))
-			client.Send(msg)
+			if err := client.Send(msg); err != nil {
+				t.Error(err)
+				done.Done()
```

Comment:
These aren't required as we're returning straight away anyway.
The test is unchanged without them.

Reply:
I was wondering about that. Do `t.Error` and `t.Errorf` halt execution?

Reply:
Actually, scrap what I said, I forgot it was in a go routine, so without using `t.Fatal` here the test won't end if there is an error.

* `t.Error` is to continue on with the rest of the test logic.
* `t.Fatal` will end a test immediately.

I'm not 100% sure which would be preferred.
I think I'd prefer to `t.Fatal` and end the test immediately if there is an error sending the message, but I can see arguments for trying to recover and finish the flow.

Reply:
I think I'm inclined to leave it alone for now, but I'm happy to change it to `t.Fatal` if @hypebeast wants.

Reply:
I'm fine with it.